Fix deleting customer emails & addresses via admin tables #7825

ashleyfae opened the pull request:
Fixes #7825

Proposed Changes:
1. Correct URL used for deleting customer email addresses.
2. Introduce function for deleting customer addresses.
3. Correct URL used for deleting customer addresses to use new function.
4. Introduce new admin notices for successfully/unsuccessfully removing addresses.
5. Add a "Delete" link when hovering over an individual address on Customer Details page. This isn't quite consistent with the Customer Emails table, which has an "Actions" column, but given the number of columns already present on the physical addresses table I though this was better.

cklosowski reviewed (dismissed):
@ashleyfae this needs to _not_ allow immediately deleting what is considered the primary email address of a customer.

When viewing the customer details, the 'primary' email address cannot be deleted, only tertiary email addresses. The same should hold true for the list table display of the 'delete' action.

cklosowski commented on includes/admin/customers/class-customer-email-addresses-table.php:
```diff
@@ -131,7 +131,14 @@ public function column_email( $item ) {
 
 		// Non-primary email actions
 		if ( 'primary' !== $item_status ) {
```
@ashleyfae I believe the issue I had with being able to delete a primary is that we're checking 'status', not 'type'.

ashleyfae commented:
@cklosowski Updated. Primary addresses now have no row action and the checkbox is disabled. I disabled the whole checkbox because our only bulk action is delete anyway.

Turns out bulk actions weren't working at all due to the nonce check failing. This has now been fixed.

Something to note related to that:

https://github.com/easydigitaldownloads/easy-digital-downloads/pull/7901/commits/18d117faed77b146e5a75d0927c16adb15456e75#diff-fc382348ba7ddea17a07f25d727c9abeL34-L35

These were being passed in via `__()` however I do not believe these arguments are intended for display/translation. They're used when creating the nonce action and in various HTML attributes (like CSS classes). Therefore we should always know exactly what they are (rather than being dynamic due to being translated). So those are now lowercase and hard-coded with that in mind.

robincornett commented on includes/admin/customers/class-customer-addresses-table.php:
```diff
@@ -31,8 +31,8 @@ class EDD_Customer_Addresses_Table extends List_Table {
 		parent::__construct( array(
-			'singular' => __( 'Address',   'easy-digital-downloads' ),
-			'plural'   => __( 'Addresses', 'easy-digital-downloads' ),
+			'singular' => 'address',
```
Are these intentionally no longer translatable strings (same for emails)? I thought these were label definitions.

ashleyfae commented:
> > Add a "Delete" link when hovering over an individual address on Customer Details page. This isn't quite consistent with the Customer Emails table, which has an "Actions" column, but given the number of columns already present on the physical addresses table I though this was better.
> 
> @ashleyfae I like the row actions instead of the columns. Should we update "Remove" action in "Customer Emails" to use the same `Delete` terminology? I think it helps convey the destruction more, especially with no confirmation.

I'm happy with that. Also if we'll be removing the `Actions` column we could use this opportunity to add a new column for `Date Added`.

spencerfinnell commented:
@ashleyfae `Date Added` sounds good too. Up to you if you want to change just the string for now and tackle the columns in the future, or if you think it makes sense to do it now.

cklosowski commented:
:+1: to adding `Date Added` if we're removing a column.

cklosowski commented:
@LisaCee do you think you can get a column added to this? @spencerfinnell, if needed can you point her in the right dirction?

ashleyfae commented:
1. Moved Customer Email actions to row actions.
2. Added new column for Date Added.
3. Fixed an undefined `customer_id` variable error when you leave the `Email Address` field blank but click `Add Email` anyway.

robincornett commented on includes/admin/customers/customers.php:
```diff
@@ -763,7 +769,7 @@ function edd_customers_view( $customer = null ) {
 				<tr>
 					<th><?php esc_html_e( 'Email',   'easy-digital-downloads' ); ?></th>
-					<th><?php esc_html_e( 'Actions', 'easy-digital-downloads' ); ?></th>
+					<th><?php esc_html_e( 'Date Added', 'easy-digital-downloads' ); ?></th>
```
The physical addresses table uses "First Used" for the column header. It feels like the data is a similar type and maybe we should use the same for each?

LisaCee replied:
I like this suggestion.  I don't care which label we use, but consistency is nice.

cklosowski replied:
I don't mind that they are different as the context is different. One is an object, the other is a property of an object than can have multiple values related to the context (email/physical address).

robincornett commented on includes/admin/customers/customers.php:
```diff
@@ -773,23 +779,25 @@ function edd_customers_view( $customer = null ) {
+							<?php else : ?>
+								<div class="row-actions">
+									<?php
+									$base_url    = admin_url( 'edit.php?post_type=download&page=edd-customers&view=overview&id=' . urlencode( $customer->id ) );
```
Not a big deal either way, but could this use `add_query_arg` instead? I find it easier to parse mentally, or actually maybe just use the `edd_get_admin_url` function to build this base URL.

ashleyfae commented on includes/admin/customers/class-customer-email-addresses-table.php:
```diff
@@ -289,10 +302,22 @@ public function process_bulk_action() {
 		}
 
+		/*
+		 * Only non-primary email addresses can be deleted, so we're building up a whitelist using the provided
```
Good catch - thank you.

robincornett commented on includes/admin/customers/customers.php:
```diff
@@ -708,6 +711,9 @@ function edd_customers_view( $customer = null ) {
 								? esc_html( $address->address2 )
 								: '';
 							?>
+							<div class="row-actions">
```
Adding the row actions does interesting things on mobile view because no cells are hidden:
![image](https://user-images.githubusercontent.com/4853066/86783363-33e69700-c02e-11ea-8d2d-5c2d4b27c29a.png)
(If the row/cell is not clicked/active, then there is just the gap.)

robincornett commented on includes/admin/customers/customers.php:
```diff
@@ -768,7 +774,7 @@ function edd_customers_view( $customer = null ) {
 				<tr>
 					<th><?php esc_html_e( 'Email',   'easy-digital-downloads' ); ?></th>
-					<th><?php esc_html_e( 'Actions', 'easy-digital-downloads' ); ?></th>
+					<th><?php esc_html_e( 'Date Added', 'easy-digital-downloads' ); ?></th>
```
The CSS for this table also needs to be updated for responsiveness, with the new column being added.

cklosowski commented:
@robincornett Do you mind taking this over so @ashleyfae can focus on the other tasks she has currently?

robincornett commented on includes/admin/customers/class-customer-email-addresses-table.php:
```diff
@@ -204,10 +211,16 @@ public function column_customer( $item ) {
+
 		return sprintf(
-			'<input type="checkbox" name="%1$s[]" value="%2$s" />',
+			'<input type="checkbox" name="%1$s[]" value="%2$s" title="%3$s"%4$s />',
```
@cklosowski I think I've addressed the rest of the PR issues, but do you want me to go ahead and fix this one here, or wait for the other PR (which would need to be updated to account for this change either way)?

spencerfinnell commented on includes/admin/customers/customers.php:
```diff
@@ -709,39 +712,42 @@ function edd_customers_view( $customer = null ) {
 							<?php
 							echo ! empty( $address->city )
-								? esc_html( $address->city )
+								? esc_attr( $address->city )
```
This should be able to stay `esc_html()`, I think? The other columns have changed as well except for Postal Code.

spencerfinnell commented on includes/admin/customers/customers.php:
```diff
@@ -777,43 +783,58 @@ function edd_customers_view( $customer = null ) {
+									$remove_url  = wp_nonce_url( add_query_arg( array( 'email' => rawurlencode( $email->email ), 'edd_action' => 'customer-remove-email'  ), $base_url ), 'edd-remove-customer-email' );
+									?>
+									<a href="<?php echo esc_url( $promote_url ); ?>"><?php esc_html_e( 'Make Primary', 'easy-digital-downloads' ); ?></a>
+									&nbsp;|&nbsp;
```
Small thing -- the extra whitespace makes these a little inconsistent with core row actions:

<img width="336" alt="Screen Shot 2020-07-10 at 1 33 16 PM" src="https://user-images.githubusercontent.com/491124/87182347-fabe5900-c2b1-11ea-8229-3b242b371259.png">

spencerfinnell commented on includes/admin/class-edd-notices.php:
```diff
@@ -660,6 +660,19 @@ private function add_user_action_notices( $notice = '' ) {
+				case 'address-removed' :
+					$this->add_notice( array(
+						'id'      => 'edd-customer-address-removed',
+						'message' => __( 'Customer address removed', 'easy-digital-downloads')
```
```suggestion
						'message' => __( 'Customer address deleted.', 'easy-digital-downloads')
```

`deleted` matches the "Delete" action link.

spencerfinnell commented on includes/admin/class-edd-notices.php:
```diff
@@ -660,6 +660,19 @@ private function add_user_action_notices( $notice = '' ) {
+				case 'address-remove-failed' :
+					$this->add_notice( array(
+						'id'      => 'edd-customer-address-remove-failed',
+						'message' => __( 'Failed to remove customer address', 'easy-digital-downloads'),
```
```suggestion
						'message' => __( 'Failed to delete customer address.', 'easy-digital-downloads'),
```

spencerfinnell commented on includes/admin/customers/customers.php:
```diff
@@ -709,39 +712,42 @@ function edd_customers_view( $customer = null ) {
 							<?php if ( ! empty( $address->is_primary ) ) : ?>
 								<span class="edd-chip"><?php esc_html_e( 'Primary', 'easy-digital-downloads' ); ?></span>
 							<?php endif; ?>
+							<div class="row-actions">
+								<span class="delete"><a href="<?php echo esc_url( $delete_url ); ?>"><?php esc_html_e( 'Delete', 'easy-digital-downloads' ); ?></a></span>
+							</div>
```
I think this makes more sense in the "Address" column. This matches core row actions always being in the first column as well as where the row actions are when viewing the "Physical Addresses" tab.

robincornett replied:
I agree with that, except for two concerns:
1. It looks/feels weird on small screens, because the row actions are after the street address and before everything else
2. Because the street address can be empty, having the row actions there presents potential accessibility issues. Generally, WP has the row actions in the first column because that's a title column, and pretty self explanatory. This is less so, because the first column isn't necessarily informative on its own, even if the street address is populated.

I ended up setting up the physical addresses table to perform somewhat like the tax rates table (or WP Core tables) for responsive behavior, but I'm not sure its data "behaves" the same way.

spencerfinnell replied:
👍 I think I'm more okay with this now. I was thinking maybe matching the global Physical Addresses table more by adding the City and Postal Code beneath the Street Address... but now I actually think doing a partially formatted version of the address like that might be more confusing than having things in separate columns (something to think about for the global table at a later date).

spencerfinnell commented on includes/admin/customers/class-customer-addresses-table.php:
```diff
@@ -135,9 +135,14 @@ public function column_address( $item ) {
 			'view' => '<a href="' . esc_url( $customer_url ) . '">' . __( 'View', 'easy-digital-downloads' ) . '</a>'
```
```suggestion
			'view' => '<a href="' . esc_url( $customer_url ) . '">' . esc_html__( 'View', 'easy-digital-downloads' ) . '</a>'
```

spencerfinnell commented on includes/admin/customers/class-customer-addresses-table.php:
```diff
@@ -135,9 +135,14 @@ public function column_address( $item ) {
+				'id'         => urlencode( $item['id'] ),
+				'edd_action' => 'customer-remove-address'
+			) ), 'edd-remove-customer-address' );
+			$actions['delete'] = '<a href="' . esc_url( $delete_url ) . '">' . __( 'Delete', 'easy-digital-downloads' ) . '</a>';
```
```suggestion
			$actions['delete'] = '<a href="' . esc_url( $delete_url ) . '">' . esc_html__( 'Delete', 'easy-digital-downloads' ) . '</a>';
```

spencerfinnell commented on includes/admin/customers/class-customer-email-addresses-table.php:
```diff
@@ -130,8 +130,15 @@ public function column_email( $item ) {
+				'email'      => rawurlencode( $email ),
+				'edd_action' => 'customer-remove-email'
+			) ), 'edd-remove-customer-email' );
+			$actions['delete'] = '<a href="' . esc_url( $delete_url ) . '">' . __( 'Delete', 'easy-digital-downloads' ) . '</a>';
```
```suggestion
			$actions['delete'] = '<a href="' . esc_url( $delete_url ) . '">' . esc_html__( 'Delete', 'easy-digital-downloads' ) . '</a>';
```
